### PR TITLE
warn user when Serf versions are out of date

### PR DESCRIPTION
**Pull request by ThorArakis**
The current behavior is to exit with an error message of "Could not find
Serf". It should be more friendly to warn if incompatible serf versions are
found in the path.

### CODE REVIEW

File: kong/cli/services/serf.lua [hunk 1]
```diff
@@ -28,7 +28,11 @@ function Serf:_get_cmd()
   local cmd, err = Serf.super._get_cmd(self, {}, function(path)
     local res, code = IO.os_execute(path.." version")
     if code == 0 then
-      return res:match("^Serf v0.7.0")
+      local version_match = res:match("^Serf v0.7.0")
+      if not version_match then
+        logger:warn("Incompatible Serf version in the path. Required version is Serf v0.7.0")
```
Comment (thibaultCha) [inline kong/cli/services/serf.lua]: Could it be more concise and to the point, like:

```
Incompatible Serf version. Kong requires v0.7.0.
```

I am not sure mentioning the `$PATH` is worthy?

File: kong/cli/services/serf.lua [hunk 2]
```diff
@@ -28,7 +28,11 @@ function Serf:_get_cmd()
...
+      end
+      return version_match
```
Comment (thibaultCha) [inline kong/cli/services/serf.lua]: Since we still execute the rest of the script here and if I compare this change with what you reported in #946, am I right to assume the output will now be this:

```
[WARN] Incompatible Serf version in the path. Required version is Serf v0.7.0
[ERR] Can't find serf
[ERR] Could not start Kong
```

If so, we should probably change the warning to be an actual error and exit the CLI, don't we?

PS: also thinking we should not actually require `0.7.0`, but `~0.7.0` or even `>=0.7.0`, but this is unrelated to this PR.

**Comment by thibaultCha** [comment]
Thanks for taking care of this!

File: kong/cli/services/serf.lua (hunk 1 above)
Comment (ThorArakis) [inline kong/cli/services/serf.lua]: Agreed. Will update.

File: kong/cli/services/serf.lua (hunk 2 above)
Comment (ThorArakis) [inline kong/cli/services/serf.lua]: I was leaving room for the possibility that you have `serf` multiple times in your path. In this case it might be possible that the first version encountered is `0.6.4` but the second one is `0.7.0`. I'm not sure how the rest of the app would work with that sort of state, but it's possible it would cause a breaking change for a small section of poorly configured users :disappointed:

Comment (ThorArakis) [inline kong/cli/services/serf.lua]: Though if you guys are ok with exiting early because of it, I'm definitely not opposed!

Comment (thibaultCha) [inline kong/cli/services/serf.lua]: I believe common practice would be to exit, indeed, and let the user re-order his or her `$PATH` if multiple Serf executables are indeed present. Otherwise, the warning, even if legit, is not that useful, and the user might not understand another version of Serf was used. Secondly, the exit error message (can't find Serf) is confusing too: it could find it, tho it just wasn't the right version.

**Comment by thibaultCha** [comment]
In order to maintain consistency across the CLI, I believe other services (such as OpenResty, Serf, but also maybe dnsmasq?) should also include the same changes as decided in this PR:

- a warning, as currently implemented
- a hard-exit, with an error message
- or at least a more descriptive error than "could not find <service name>" (since the actual error is a version mismatch).

Thoughts? @thefosk as he built the current services management.

**Comment by thibaultCha** [comment]
To clear up the fog here, we need to do one of these two things:

- Either exist as soon as an incompatible version is detected, let the user deal with his `$PATH` ordering
- Wait until all possible `serf` binaries have been tested before exiting with an error.

And this behavior needs to be reproduced for each sub-service (OpenResty, dnsmasq if used, Serf). @thefosk thoughts?

I believe we can already implement option 1 for this PR since @ThorArakis already started working on it and it was the original scope of this PR.

**Comment by ThorArakis** [comment]
@thibaultCha updated serf check to log as an error and exit immediately and then added similar check for nginx. Looks like dnsmasq doesn't do version matching so I didn't modify it.

**Comment by Tieske** [comment]
closing this in favour of #1221
